### DISCUSSION TITLE
Groups based on LDAP groups

### DISCUSSION THREAD

**Discussion by q-wertz** [requarks/wiki#6139] [Help / Questions] [answered]
Thanks for the really nice new feature added in https://github.com/requarks/wiki/pull/5903 :blush: A few questions came up when setting it up and testing things out:

1. It looks like existing Authentication Strategies cannot be changed to also map the groups, so I created a new one (where these settings are available). But now the same user in LDAP exists twice on Wiki (from both Providers, with different User-`ID`). Is there a way to circumvent this?
2. After removing the user from the AD group which allows the login and of some others, he cannot login after logout (which is good). But in the Wiki admin management user overview
    1. he is still shown as `Active` 
    2. he is still assigned to the groups
    
    Does this reset/update after a certain time?\
    And will he be logged out by the server after a certain time?
4. A [code comment](https://github.com/requarks/wiki/commit/8fa771c4ce6b667d5c2b1de6b343977af4351644#diff-e62bba4f65b928357a9c4a9fbc4f794c486677f609f1ea3945bb56937a7f937eR50) states, that 
    > remove any groups that don't match LDAP

    On our server there is one group that is not available on the AD server, but it didn't get deleted. Did I misinterpret the comment? (Not too familiar with js)
    **EDIT:**
    As far as I understand the js code the group mapping gets removed not the whole group. Sorry, misunderstanding of the comment from my side.
5. It looks like nested groups are not supported by WikiJS?
    **EDIT:**
    My fault, one has to change the `Group Search Filter` to `(member:1.2.840.113556.1.4.1941:={{dn}})`

**Comment by q-wertz** [comment]
Is there information on 1. and 2. whether this is by design, a "bug" or unfinished implementation?

**Reply by NGPixel** [reply]
1. You should be able to enable group mapping on existing strategies. If not, this is a bug. As for creating a new strategy, this will create issues as users can only be tied to a single strategy. This has been addressed in v3 where users can have any number of strategies tied to them.
2. This is as designed. There's no "sync" occurring between Wiki.js and your LDAP directory. The group mappings only get updated upon login.

**Reply by q-wertz** [reply]
Thanks for the info. :blush: 

1. As somebody else https://github.com/requarks/wiki/discussions/6158 described the same problem it might really be a bug. Thanks for the info of the changes in v3.
2. Ok :+1:

**Reply by q-wertz** [reply]
Another one https://github.com/requarks/wiki/discussions/6181

**Reply by q-wertz** [reply]
@icsinfo as you are the author of the PR could you maybe shine some light onto question no. 1 (the group mapping settings are not available for existing authentication strategies, only new ones)? 

Btw thanks for this really nice feature :blush: :rocket:

**Comment by icsinfo** [comment]
I did notice this with testing it, and was not able to figure out how to make it available for current auth strategies. My guess is because the json string field in the database that defines the options for the existing ldap strategy do not contain the new options, they won’t show. When you make a new strategy, the json string field is populated with the new options, this is why they show. We will probably have to cook up some sort of migration for existing configurations that inserts the new options if they don’t already exist. 

In the interim, if you configure a second ldap auth strategy with all the necessary options and test it, you *should* (haven’t tested this, going to try it this week) be able to update the existing auth strategy via sql query with the string from the new one. Or at the very least, take your current json string, and append the new options to it. Then you should be able to define them in the admin interface.

**Reply by q-wertz** [reply]
Thanks for your help. Depending on how much time I have this week I will try out one of the options and report back :+1:

**Comment by icsinfo** [comment]
Yes, that is exactly what the issue is. Pre group mapping, the json in the config column of the authentication table is...

```json
{
    "url": "ldap://serverhost:389",
    "bindDn": "cn='root'",
    "bindCredentials": "",
    "searchBase": "o=users,o=example.com",
    "searchFilter": "(uid={{username}})",
    "tlsEnabled": false,
    "verifyTLSCertificate": true,
    "tlsCertPath": "",
    "mappingUID": "uid",
    "mappingEmail": "mail",
    "mappingDisplayName": "displayName",
    "mappingPicture": "jpegPhoto"
}
```
Configuring a new ldap strategy after the group mapping update adds the new fields.

```json
{
    "url": "ldap://serverhost:389",
    "bindDn": "cn='root'",
    "bindCredentials": "",
    "searchBase": "o=users,o=example.com",
    "searchFilter": "(uid={{username}})",
    "tlsEnabled": false,
    "verifyTLSCertificate": true,
    "tlsCertPath": "",
    "mappingUID": "uid",
    "mappingEmail": "mail",
    "mappingDisplayName": "displayName",
    "mappingPicture": "jpegPhoto",
    "mapGroups": false,
    "groupSearchBase": "OU=groups,dc=example,dc=com",
    "groupSearchFilter": "(member={{dn}})",
    "groupSearchScope": "sub",
    "groupDnProperty": "dn",
    "groupNameField": "name"
}
```

I was able to use the following query in my test environment to add the options to an existing ldap auth strategy (we use postgres). This will just append the defaults with mapping groups disabled and *shouldn't* affect how your strategy works currently. You will then see the options in the front end to configure.  The standard use at your own risk, it may bork your install disclaimer applies.
```sql
update authentication
set config = config :: jsonb || '{"mapGroups": false,
    "groupSearchBase": "OU=groups,dc=example,dc=com",
    "groupSearchFilter": "(member={{dn}})",
    "groupSearchScope": "sub",
    "groupDnProperty": "dn",
    "groupNameField": "name"}'
    where key = 'unique key for ldap auth strategy'
```
@NGPixel Lines 101 - 106 of models\authentication.js seem to try to insert the new properties on an auth strategy that hasn't been updated, but I don't see where it writes it back to the database. Perhaps this is where the bug lies?
```js
        strategy.config = _.transform(strategyDef.props, (result, value, key) => {
          if (!_.has(result, key)) {
            _.set(result, key, value.default)
          }
          return result
        }, strategy.config)
```

**Reply by q-wertz** [reply]
Also manually updated the entry in the database. Works like a charm now :blush:

**Comment by q-wertz** [comment, marked as answer]
Looks like this was fixed in [v2.5.298](https://github.com/requarks/wiki/releases/tag/v2.5.298) by @icsinfo in 5acc7e752e836fbe771e829e3e7218d8438e3256

Thanks :blush:
